core/src/impl/Kokkos_Profiling.cpp

```diff
@@ -114,27 +113,27 @@ void beginParallelFor(const std::string& kernelPrefix, const uint32_t devID,
     Kokkos::fence();
     (*Experimental::current_callbacks.begin_parallel_for)(kernelPrefix.c_str(),
                                                           devID, kernelID);
+  }
 #ifdef KOKKOS_ENABLE_TUNING
-    auto context_id = Experimental::get_new_context_id();
-    Experimental::begin_context(context_id);
-    Experimental::VariableValue contextValues[] = {
-        Experimental::make_variable_value(
-            Experimental::kernel_name_context_variable_id, kernelPrefix),
-        Experimental::make_variable_value(
-            Experimental::kernel_type_context_variable_id, "parallel_for")};
-    Experimental::set_input_values(context_id, 2, contextValues);
+  auto context_id = Experimental::get_new_context_id();
```

Comment:
I guess because its internal tuning which is controlled via KOKKOS_ENABLE_TUNING not via some runtime parameter. We may wanna have a runtime parameter nevertheless (like --kokkos-enable-tuning=1/0 or so)

Reply:
Actually, this is _really_ subtle, and it's a thing people should know. You can absolutely tune without having a begin_parallel_for callback set. In fact, the way the prototype tool works, you _learn_ using a tool that has the begin/end parallel_x callbacks (and thus get fences and nice timings), but _apply_ using a tool that doesn't have them set (and thus doesn't insert fences to hose your performance).

I don't really like how subtle this thing is in some places. I think @crtrott 's comment about having a runtime flag to enable/disable tuning is a good one, which would remove some of the hidden behavior here...